src/N3Store.js

```diff
@@ -1039,8 +1059,27 @@ class DatasetCoreAndReadableStream extends Readable {
     if (!this._filtered) {
       const { n3Store, graph, object, predicate, subject } = this;
       const newStore = this._filtered = new N3Store({ factory: n3Store._factory, entityIndex: this.options.entityIndex });
-      for (const quad of n3Store.readQuads(subject, predicate, object, graph))
-        newStore.addQuad(quad);
+      const graphs = n3Store._getGraphs(graph);
+
+      let subjectId, predicateId, objectId;
+
+      // Translate IRIs to internal index keys.
+      if (subject   && !(subjectId   = newStore._termToNumericId(subject))   ||
+          predicate && !(predicateId = newStore._termToNumericId(predicate)) ||
+          object    && !(objectId    = newStore._termToNumericId(object)))
+        return newStore;
+
+      for (const graph in graphs) {
+        const subjects = indexMatch(graphs[graph].subjects, [subjectId, predicateId, objectId]);
+        if (subjects) {
+          newStore._graphs[graph] = {
+            subjects,
+            predicates: indexMatch(graphs[graph].predicates, [predicateId, objectId, subjectId]),
+            objects: indexMatch(graphs[graph].objects, [objectId, subjectId, predicateId]),
```

Comment:
We could optimise slightly further by using the same technique as `#match` to determine which index to match over first as this will be more efficient for identifying if there are no matches. But it would be sufficient to leave that work for a follow up PR.